cranelift/codegen/src/isa/aarch64/abi.rs

```diff
@@ -1525,16 +1520,99 @@ fn create_reg_env(enable_pinned_reg: bool) -> MachineEnv {
                 preg(vreg(15)),
             ],
             // Vector Regclass is unused
-            vec![],
+            &[],
         ],
-        fixed_stack_slots: vec![],
+        fixed_stack_slots: &[],
         scratch_by_class: [None, None, None],
-    };
-
-    if !enable_pinned_reg {
-        debug_assert_eq!(PINNED_REG, 21); // We assumed this above in hardcoded reg list.
-        env.non_preferred_regs_by_class[0].push(preg(xreg(PINNED_REG)));
     }
+};
 
-    env
-}
+static DEFAULT_MACHINE_ENV: MachineEnv = {
+    debug_assert!(PINNED_REG == 21); // We assumed this below in hardcoded reg list.
```

Comment:
This probably goes in the `PINNED_MACHINE_ENV` above, no? (It's load-bearing there that we get it right, because we *excluded* the pinned register from allocation; the default environment is used when pinning is disabled so doesn't care what the pinned reg is)